File: library/CreateCovars.js
const fs = require('fs');
let linear = require('./everpolate').linear;

let create_covars = function(dir_covidtesting,endTime="2020-04-27",predTime=null){
	let t0 = new Date("2019-12-31");
	let tf = new Date(endTime);

	// read all rows and chaeck the border time and convert the selected colnames
	let temp, file;
	file = fs.readFileSync(dir_covidtesting).toString();
	let lines = file.split(/\r\n|\n/);
	let selected_colnames = ["Reported Date",
		"Total patients approved for testing as of Reporting Date",
		"Under Investigation"];
	let dataCovar_Index = [];	
	let dataCovar = [];
	let dataCovar_temp = lines[0].replace(/['"]+/g, '').split(',');
	for(let i = 0; i < selected_colnames.length; i++){
		dataCovar_Index.push(dataCovar_temp.indexOf(selected_colnames[i]));
	}
  let startdate = new Date("2020-02-03");
	for (let i = 1; i < lines.length; i++) {
		let temp = lines[i].split(',');
		if(temp.length > 1) {
			tempDate = new Date(temp[dataCovar_Index[0]]);
			if( tempDate > startdate && tempDate <= tf){
				let tempDayCount = Math.ceil(Math.abs(tempDate - t0) / (1000 * 3600 * 24));
			  let tempcovar =	[tempDayCount];
				for(let j = 1; j < selected_colnames.length; j++){
					tempcovar.push(Number(temp[dataCovar_Index[j]]));
				}
				dataCovar.push(tempcovar);
			}
		}
	}

	let coaverTimes = dataCovar.map(x => x[0]);
	let coaverData = dataCovar.map(x => x[1]);
	let timeLength = dataCovar[dataCovar.length-1][dataCovar_Index[0]];
	let time = Array.from(Array(timeLength), (_, i) => i + 1);

	let total_tests = linear(time, coaverTimes, coaverData);
	total_tests = total_tests.map(x => Math.ceil(x));
  
  tests = total_tests.map( (_,i,arr) => Math.max((arr[i]) - (arr[i-1]),0));
	tests = tests.map( (_,i,arr) => Math.ceil(arr[i] * 0.5 + arr[i+1] * 0.5));
	tests[timeLength - 1] = tests[timeLength - 2];

	let data = time.map( (x,i) => [x,tests[i]]);
	data.unshift(["time","tests"]);

	if (predTime != null) {
    predTime = new Date(predTime);
		let add_time = Math.ceil(Math.abs(tf - predTime) / (1000 * 3600 * 24));
		for(let i =0; i < add_time; i++)
			data.push([timeLength + i,10e3]);
	}
	
	return data;
}

module.exports = create_covars;

File: R-forward-function/CreateCovars.R

create_covars <- function(endTime="2020-09-07",predTime=NULL) {
  
  t0 <- as.Date("2019-12-31")
  tf <- as.Date(endTime)

  dat <- read.csv("covidtesting.csv", header=TRUE)
  dat$Reported.Date <- as.Date(dat$Reported.Date)
  dat <- subset(dat, Reported.Date>as.Date("2020-02-03") & Reported.Date<=tf)
  dat$Reported.Date <- as.numeric(as.Date(dat$Reported.Date))-as.numeric(t0)
  dat <- subset(dat, select=c("Reported.Date",
                              "Total.patients.approved.for.testing.as.of.Reporting.Date",
                              "Under.Investigation"))
  colnames(dat) <- c("Date","TotalTests","Pending")

  time <- seq(1,dat$Date[length(dat$Date)])
  total_tests <- approx(dat$Date,dat$TotalTests,xout=time,method="linear")$y
  total_tests <- ceiling(total_tests)
  
  
  pending <- rep(0,length(time))
  pending[dat$Date] <- dat$Pending 
  
  tests <- rep(NA,length(time))
  
  for (i in 2:length(time)) {
    tests[i] <- max((total_tests[i]) - (total_tests[i-1]),0)
  }
  tests <- ceiling(filter(tests, rep(1/2,2)) )
  tests[length(tests)] <- tests[length(tests)-1]
  
  
  
  data <- data.frame(time,ceiling(as.numeric(tests)))
  colnames(data) <- c("time","tests")
  
  
  
  
  
  if (!is.null(predTime)) {
    predTime <- as.Date(predTime)
    add_time <- seq(as.numeric(tf)+1,as.numeric(as.Date(predTime)))-as.numeric(t0)
    add_tests <- rep(10e3,length(add_time))
    add_data <- cbind(add_time,add_tests)
    colnames(add_data) <- c("time","tests")
    data <- rbind(data,add_data)
    data <- as.data.frame(data)
    
  }
  
  return(data)
}
